internal/searchindex/searchindex.go

```diff
@@ -0,0 +1,440 @@
+			}
+
+			err := compat.JSONCopy(&res, *in)
+			return res, err
+		}
+
+		errs := []error{}
+		for i := range *in {
+			tokenFilters, err := convertFilters((*in)[i].TokenFilters)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("unable to convert tokenFilters: %w", err))
+				continue
+			}
+
+			charFilters, err := convertFilters((*in)[i].CharFilters)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("unable to convert charFilters: %w", err))
+				continue
+			}
+
+			tokenizer, err := convertTokenizer(&(*in)[i].Tokenizer)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("unable to convert tokenizer: %w", err))
+				continue
+			}
+
+			result = append(result, akov2.AtlasSearchIndexAnalyzer{
+				Name:         (*in)[i].Name,
+				TokenFilters: tokenFilters,
+				CharFilters:  charFilters,
+				Tokenizer:    tokenizer,
+			})
+		}
+		e := errors.Join(errs...)
+		return &result, e
+	}
+
+	convertStoredSource := func(in map[string]interface{}) (*apiextensionsv1.JSON, error) {
+		val, err := json.Marshal(in)
+		if err != nil {
+			return nil, err
+		}
+		return &apiextensionsv1.JSON{Raw: val}, nil
+	}
+
+	var errs []error
+	analyzers, err := convertAnalyzers(index.Analyzers)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	storedSource, err := convertStoredSource(index.StoredSource)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	return &SearchIndex{
+		SearchIndex: akov2.SearchIndex{
+			Name:           index.Name,
+			DBName:         index.Database,
+			CollectionName: index.CollectionName,
+			Type:           pointer.GetOrDefault(index.Type, ""),
+			Search:         search,
+			VectorSearch:   vectorSearch,
+		},
+		AtlasSearchIndexConfigSpec: akov2.AtlasSearchIndexConfigSpec{
+			Analyzer:       index.Analyzer,
+			Analyzers:      analyzers,
+			SearchAnalyzer: index.SearchAnalyzer,
+			StoredSource:   storedSource,
+		},
+		ID:     index.IndexID,
+		Status: index.Status,
+	}, errors.Join(errs...)
+}
+
+func (s *SearchIndex) cleanup(cleaners ...indexCleaner) {
+	s.ID = nil
+	s.Status = nil
+
+	for _, cleaner := range cleaners {
+		cleaner(s)
+	}
+}
+
+type indexCleaner func(s *SearchIndex)
+
+func cleanConfigRef() indexCleaner {
+	return func(s *SearchIndex) {
+		if s.Search == nil {
+			return
+		}
+		s.Search.SearchConfigurationRef.Name = ""
+		s.Search.SearchConfigurationRef.Namespace = ""
+	}
+}
+
+// This is because API never returns nil
+func cleanVectorSearch() indexCleaner {
+	return func(s *SearchIndex) {
+		if s.VectorSearch == nil {
+			s.VectorSearch = &akov2.VectorSearch{}
+		}
+	}
+}
+
+// This is because API never returns nil
+func cleanSynonyms() indexCleaner {
+	return func(s *SearchIndex) {
+		if s.Search == nil {
+			return
+		}
+		if s.Search.Synonyms == nil {
+			s.Search.Synonyms = &([]akov2.Synonym{})
+		}
+	}
+}
+
+func cleanStoredSource() indexCleaner {
+	return func(s *SearchIndex) {
+		if s.StoredSource == nil {
+			return
+		}
+		if string(s.StoredSource.Raw) == "null" {
```

Comment:
please make this explicit as a comment why we need to handle it, additionally please create a ticket to discuss with the API-X team.

Reply:
Added comment to the godoc